notes/memory-and-management.md

```diff
@@ -0,0 +1,666 @@
+    delete[] sameFirst;
+    delete[] sameSecond;
+    return 0;
+}
+```
+
+Answer: Note the relationship between `sameFirst` and `sameSecond`. They are physically the same block of memory. It is allocated into `arrayFirst`, modifed inside of `FunctionFirst()`, returned into `sameFirst`, then fed into `FunctionSecond()` as `arraySecond`, modified, and returned as `sameSecond`. The reason this is a problem is that `sameFirst` and `sameSecond` are deleted, but they refer to the same block of memory, which means the code attempts to delete the array twice, and this is the error.
+
+Solution 1: Delete either `sameFirst` or `sameSecond`, just not both.
+
+Solution 2: In `FunctionSecond()`, instead equating `arr` and `arraySecond`, initialize `arraySecond` properly with `malloc` or something else, and in the for loop, set each element to one more than the corresponding element in `arr`.
+
+Solution 3: Just don't even make a `sameSecond` or `FunctionSecond()` and go straight to the answer.
+
+---
+
+The error in Problem 2 is known as a **double free** and can occur if you aren't being careful about what memory is which.
+
+---
+Problem 4:
+
+Below is a modifed version of the code from Problem 3. What is wrong with it?
+
+```C++
+#include <stdlib.h>
+#include <iostream>
+
+int* FunctionFirst(int size) {
+    int* arrayFirst = (int *) calloc(sizeof(int) * size);
+    arrayFirst[0] = 1;
+    for(int i = 1; i < size; i++) {
+        arrayFirst[i] += 1 + arrayFirst[i - 1];
+    }
+    return arrayFirst;
+}
+
+double* FunctionSecond(int* arr, int size) {
+    double* arraySecond = (double *) arr
+    for(int i = 0; i < size; i++) {
+        arraySecond[i] += 1.0;
+    }
+    return arraySecond;
+}
+
+int main() {
+    int* sameFirst = FunctionFirst(400);
+    double* sameSecond = FunctionSecond(sameFirst, 400);
+    std::cout << "Sum from 1 to 400 as a Double plus 1: " << (double) sameSecond(sameSecond[399]);
+    delete[] sameFirst;
+    return 0;
+}
+```
+
+Answer: Note the cast in done to produce `arraySecond`. Casting in a language that lacks type-safety is risky, but it is fatal in this example. A cast from an int to double array reinterprets the array as as one with blocks of 8 bytes, so `arraySecond` has not 400 ints, but 200 doubles! When the for loop in `FunctionSecond` triggers, it will access an element out of bounds and will cause an error.
+
+Solution 1: Iterate through `arr` and assign it to an already initialized `arraySecond` that has 400 * 8 = 3200 bits. 
+
+Solution 2: Use a valid casting operator, found in the STL library. Just as long as its not std::reinterpret_cast<T*>, please.
+
+Solution 3: Just don't, this is a bad thing to do in general.
+
+---
+
+This is an error so special to C++ that you don't see it in other languages. This is called a **Segmentation Fault**, and occurs because the program attempted to reach memory it wasn't allowed to reach.
+
+## The Stack
```

Comment:
Totally up to you, but I'd suggest somehow telling the reader that the "heap" and "stack" sections are by far the most important, since they directly impact the programmer